content/struct-classes.md

```diff
@@ -0,0 +1,502 @@
+methods to set the value of the `email` field:
+
+~~~ scala
+// 1st solution, as above
+def withEmail(newEmail: Option[String]): User
+// 2nd solution
+def withEmail(newEmail: String): User
+def withoutEmail: User
+~~~
+
+Here, in the 2nd solution, calling `withEmail` would always set the email to
+`Some(newEmail)`, and calling `withoutEmail` would set it to `None`. We can imagine
+similar convenience methods for fields containing collections of values.
+
+Internally, a `struct class` is implemented in a similar way as a `case class`: the
+compiler synthesizes the `private` method `copy`, and overrides the default implementations
+of `toString`, `hashCode`, and `equals`. For reference, here is what the definition
+of the first version of `User` "desugars" to:
+
+~~~ scala
+import scala.util.hashing.MurmurHash3
+
+class User(val name: String, val age: Int) extends Serializable:
+  private def copy(name: String = name, age: Int = age): User = User(name, age)
+  override def toString() = s"User($name, $age)"
+  override def hashCode() =
+    MurmurHash3.finalizeHash(
+      MurmurHash3.mix(MurmurHash3.mix(MurmurHash3.productSeed, name), age),
+      2
+    )
+  override def equals(that: Any): Boolean =
+    that match
+      case user: User => user.name == name && user.age == age
+      case _ => false
+
+  def withName(newName: String): User = copy(name = newName)
+  def withAge(newAge: Int): User = copy(age = newAge)
+end User
+~~~
+
+### Specification
+
+A struct class is a `class` definition prefixed with `struct`:
+
+~~~ text
+TmplDef ::= 'struct' 'class' ClassDef
+~~~
+
+A struct class is required to have a parameter list that is not contextual. The
+formal parameters in the first parameter list are called _elements_ and are treated
+specially. A `val` prefix is implicitly added to such a parameter, unless the parameter
+already carries a `val` or `var` modifier. Hence, an accessor definition for the
+parameter is generated.
+
+A struct class definition `c[tps](ps_1)...(ps_n)` with type parameters `tps` and
+value parameters `ps` is handled as follows. A `private` method named `copy` is
+implicitly added to the class definition unless the class already has a member
+(directly defined or inherited) with that name, or the class has a repeated parameter.
+The method is defined as follows:
+
+~~~ scala
+private def copy[Ts](ps_1')...(ps_n'): c[Ts] = c[Ts](xs_1)...(xs_n)
+~~~
+
+Where `Ts` is the vector of types defined in the type parameter section of the class
+definition. Each `xs_i` denotes the parameter names of the parameter section `ps_i`.
+The value parameters `ps_1j'` of the first parameter list have the form
+`x_1j: T_1j = this.x_1j`. The other parameters `ps_ij'` of the `copy` method are
+defined as `x_ij: T_ij`. In all cases, `x_ij` and `T_ij` refer to the name and type
+of the corresponding class parameter `ps_ij`.
+
+Every struct class implicitly overrides some method definitions of class `scala.AnyRef`
```

Comment:
Shall struct classes be implicitly `final`?

Reply:
That’s a good question. I am tempted to stay conservative and follow the same behavior as case classes (ie, forbid struct-to-struct inheritance, struct-to-case inheritance, and case-to-struct inheritance, but not make struct classes `final`). 

That being said, if we allow a class to extend a struct class, we should probably make struct classes extend [Equals](https://www.scala-lang.org/api/current/scala/Equals.html) and use its `canEqual` method to implement `equals` to make sure we don’t generate an implementation of `equals` that is not an equivalence relation.

Reply:
My take would be that these should be final. What would be the usecase for subclassing these?